### ISSUE TITLE
openssl x509 returns 0 even though the date of certificate is invalid

### ISSUE THREAD

**Issue by SuhwanSong** [openssl/openssl#11772]
### Description
There is an invalid behavior when verifying the validity of expiration date of certificate.
openssl returns 0 even though the date of certificate (pem file) has expired.
version : openssl-1.0.2u

### How to reproduce
1. build openssl-1.0.2u
2. unzip PoC.zip
3. run following commands
`openssl x509 -in $PoC -checkend 0 -noout`
`openssl x509 -in $PoC -text -noout`

PoC : https://github.com/openssl/openssl/files/4600224/PoC.zip
[PoC.zip](https://github.com/openssl/openssl/files/4600224/PoC.zip)

Here's log
```
$ openssl x509 -in $PoC -checkend 0 -noout
exit status : 0
```

```
$ openssl x509 -in $PoC -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: -24 (-0x18)
    Signature Algorithm: md4WithRSAEncryption
        Issuer: C=CN, ST=PN, L=CN, O=ON, OU=UN, CN=L\xA5rong Yang
        Validity
            Not Before: Jul 15 21:19:00 2005 GMT
            Not After : Aug 14 21:19:47 2005
        Subject: C=CN, ST=PN, L=CN/2.8756.10=ON, OU=UN, CN=Herong Yang
        Subject Public Key Info:
            Public Key Algorithm: md4WithRSAEncryption
            Unable to load Public Key
140718862161600:error:0609E09C:digital envelope routines:PKEY_SET_TYPE:unsupported algorithm:p_lib.c:239:
140718862161600:error:0B07706F:x509 certificate routines:X509_PUBKEY_get:unsupported algorithm:x_pubkey.c:148:
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                55:C8:EF:49:2B:5D:E4:03:C5:98:1B:68:24:28:47:88:D4:0F:D7:04
            X509v3 Authority Key Identifier:
                0v..U..I+].....h$(G...w..[.Y0W1.0...U....CN1.0...U....PN1w0...U....CN1.0...U.
..ON1.0...U....UN1.0...U....Herong Yang...
            X509v3 Basic Constraints:
                CA:TRUE
    Signature Algorithm: 2.18.840.113549.1.1.4
         3f:ba:0c:c1:ae:38:ca:f6:37:16:9c:35:5f:18:79:64:de:27:
         0d:46:ad:28:57:b4:62:df:6c:ff:f8:74:17:eb:b6:91:7e:06:
         6e:ec:a5:9d:23:e2:6e:5a:6e:c3:09:fa:cf:34:65:70:15:65:
         10:3a:6b:0f:b9:ef:6b:64:18:0e
```

Environment :
- Description: Ubuntu 18.04.1 LTS
- Release: 18.04

**Comment by kroeckx** [comment]
The problem is that the date format is invalid. The manual page
documents this as:

BUGS

       Unlike many standard comparison functions, X509_cmp_time()
       and X509_cmp_current_time() return 0 on error.

The problem is that the function calling X509_cmp_time() doesn't
check that it returned an error.

PS: You seem to be fuzzing. I suggest you fuzz in der format, not
in pem format.

**Comment by SuhwanSong** [comment]
Then, isn't this kind of issue a bug?

Also, in the case of openssl 1.1.1g, it returns 1
```
openssl x509 -in ./id:000001 -checkend 0 -noout
unable to load certificate
140194777551296:error:0908F070:PEM routines:get_header_and_data:short header:../crypto/pem/pem_lib.c:812:
exit status : 1
```

Thanks for your suggestion!

**Comment by kroeckx** [comment]
> Then, isn't this kind of issue a bug?

Yes, like I said, x509 should check the return value. It now does:

        if (X509_cmp_time(X509_get0_notAfter(x), &tcheck) < 0) {
            BIO_printf(out, "Certificate will expire\n");
            ret = 1;
        } else {
            BIO_printf(out, "Certificate will not expire\n");
            ret = 0;
        }

So if it returns 0 (error), it will just say it will not expire.


> openssl x509 -in ./id:000001 -checkend 0 -noout
> unable to load certificate
> 140194777551296:error:0908F070:PEM routines:get_header_and_data:short header:../crypto/pem/pem_lib.c:812:
> exit status : 1

Yes, that's because it's invalid PEM that 1.0.2 seems to accept.
If you convert it to der with an old version, then try to read it
in a new version, it will print:
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: -24 (-0x18)
        Signature Algorithm: md4WithRSAEncryption
        Issuer: C = CN, ST = PN, L = CN, O = ON, OU = UN, CN = L\C2\A5rong Yang
        Validity
            Not Before: Bad time value

And with -checkend it will also print the "Certificate will not
expire" message.

If you check out the repository, there is a large corpus of x509
certificates in der format in fuzz/corpora/x509

**Comment by mattcaswell** [comment]
> Then, isn't this kind of issue a bug?

Just to add to @kroeckx's comments - yes it is a bug. However the 1.0.2 version is no longer supported.

**Comment by t8m** [comment]
And the issue is fixed in all supported branches.

**Comment by kroeckx** [comment]
That the return value is ignore doesn't seem to be fixed in any branch.

Note that the supported branches gives an additional error, it
doesn't get far enough.

**Comment by t8m** [comment]
I am sorry for mistakenly closing the report then. I misunderstood that there are actually two issues on 1.0.2 branch of which only one is fixed in active branches.

**Comment by t8m** [comment]
ping @n13l This should be an easy backlog fix
